### PR TITLE
NEXUS-21897: P2 Metadata self-repair

### PR DESCRIPTION
**Pull request by DreamVan**
[Jira NEXUS-21897](https://issues.sonatype.org/browse/NEXUS-21897)

### CODE REVIEW

File: nexus-repository-p2/src/test/java/org/sonatype/nexus/repository/p2/internal/proxy/P2ProxyFacetImplTest.java [hunk 1]
```diff
@@ -59,28 +59,28 @@
...
 
   @Mock
   private TempBlob tempBlob;
 
-  private P2ProxyFacetImpl underTest;
-
   @Before
   public void setUp() {
-    underTest = new P2ProxyFacetImpl(p2DataAccess, artifactsXmlAbsoluteUrlRemover, xmlParser, jarParser);
+    p2DataAccess = new P2DataAccess(xmlParser, manifestParser);
```
Comment (sergeyonofreychuck) [inline nexus-repository-p2/src/test/java/org/sonatype/nexus/repository/p2/internal/proxy/P2ProxyFacetImplTest.java]: why testing of p2DataAccess in P2ProxyFacetImplTest.java  ?

File: nexus-repository-p2-it/src/test/java/org/sonatype/nexus/repository/p2/internal/P2RestoreBlobIT.java [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+import static org.sonatype.nexus.repository.storage.Bucket.REPO_NAME_HEADER;
+
+public class P2RestoreBlobIT
+    extends P2ITSupport
+{
+  @Inject
+  private BlobstoreRestoreTestHelper testHelper;
+
+  @Inject
+  @Named("p2")
```
Comment (proteus1121) [inline nexus-repository-p2-it/src/test/java/org/sonatype/nexus/repository/p2/internal/P2RestoreBlobIT.java]: P2Format.NAME if possible

File: nexus-repository-p2-it/src/test/java/org/sonatype/nexus/repository/p2/internal/P2RestoreBlobIT.java [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+  private RestoreBlobStrategy p2RestoreBlobStrategy;
+
+  private static final String PROXY_REPO_NAME = "p2-proxy";
```
Comment (proteus1121) [inline nexus-repository-p2-it/src/test/java/org/sonatype/nexus/repository/p2/internal/P2RestoreBlobIT.java]: P2ProxyRecipe.NAME if possible

File: nexus-repository-p2/src/main/java/org/sonatype/nexus/repository/p2/internal/util/P2DataAccess.java [hunk 4]
```diff
@@ -13,124 +13,62 @@
...
-   *
-   * @return blob content
-   */
-  public Content saveAsset(final StorageTx tx,
-                           final Asset asset,
-                           final Supplier<InputStream> contentSupplier,
-                           final String contentType,
-                           @Nullable final AttributesMap contentAttributes) throws IOException
+  @VisibleForTesting
+  public P2Attributes mergeAttributesFromTempBlob(final TempBlob tempBlob, final P2Attributes sourceP2Attributes)
```
Comment (sergeyonofreychuck) [inline nexus-repository-p2/src/main/java/org/sonatype/nexus/repository/p2/internal/util/P2DataAccess.java]: why P2DataAccess, (should be something related to DB from the name) has only one function mergeAttributesFromTempBlob

File: nexus-repository-p2/src/main/java/org/sonatype/nexus/repository/p2/internal/P2RestoreFacetImpl.java [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+  {
+    TempBlob tempBlob = getTempBlob(blob, blobStoreName);
+
+    Map<String, String> attributes = new HashMap<>();
+    AssetKind assetKind = facet(P2Facet.class).getAssetKind(blobName);
+    if (AssetKind.COMPONENT_BINARY.equals(assetKind)) {
+      //https/download.eclipse.org/technology/epp/packages/2019-12/binary/epp.package.java.executable.cocoa.macosx.x86_64_4.14.0.20191212-1200
+      String[] versionPaths = blobName.split("_");
+      String version = versionPaths[versionPaths.length - 1];
+      String[] namePaths = blobName.split("/");
```
Comment (proteus1121) [inline nexus-repository-p2/src/main/java/org/sonatype/nexus/repository/p2/internal/P2RestoreFacetImpl.java]: P2PathUtils.DIVIDER

File: nexus-repository-p2/src/main/java/org/sonatype/nexus/repository/p2/internal/P2FacetImpl.java [hunk 6]
```diff
@@ -0,0 +1,283 @@
...
+ */
+@Named
+public class P2FacetImpl
+    extends FacetSupport
+    implements P2Facet
+{
+  public static final List<HashAlgorithm> HASH_ALGORITHMS = ImmutableList.of(SHA1);
+
+  @Override
+  public Component findOrCreateComponent(final StorageTx tx, final String path, final Map<String, String> attributes) {
```
Comment (sergeyonofreychuck) [inline nexus-repository-p2/src/main/java/org/sonatype/nexus/repository/p2/internal/P2FacetImpl.java]: the path argument not used

File: nexus-repository-p2/src/main/java/org/sonatype/nexus/repository/p2/internal/util/P2DataAccess.java [hunk 7]
```diff
@@ -13,124 +13,62 @@
...
+import static com.google.common.base.Preconditions.checkNotNull;
 
 /**
  * Shared code between P2 facets.
  */
 @Named
 public class P2DataAccess
 {
-  public static final List<HashAlgorithm> HASH_ALGORITHMS = ImmutableList.of(SHA1);
+  final Logger log = Loggers.getLogger(this);
```
Comment (proteus1121) [inline nexus-repository-p2/src/main/java/org/sonatype/nexus/repository/p2/internal/util/P2DataAccess.java]: As far as I know you can extend this class from ComponentSupport and use logger from parent

File: nexus-repository-p2/src/main/java/org/sonatype/nexus/repository/p2/internal/P2RestoreFacetImpl.java [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+    BlobStore blobStore = checkNotNull(blobStoreManager.get(blobStoreName));
+    InputStream inputStream = blob.getInputStream();
+    MultiHashingInputStream hashingStream = new MultiHashingInputStream(HASH_ALGORITHMS, inputStream);
+    return new TempBlob(blob, hashingStream.hashes(), true, blobStore);
+  }
+
+  private Map<String, String> getComponentAttributes(final Blob blob, final String blobName, final String blobStoreName)
+      throws IOException
+  {
+    TempBlob tempBlob = getTempBlob(blob, blobStoreName);
```
Comment (sergeyonofreychuck) [inline nexus-repository-p2/src/main/java/org/sonatype/nexus/repository/p2/internal/P2RestoreFacetImpl.java]: TempBlob is closeble. Use "try ("

File: nexus-restore-p2/src/main/java/org/sonatype/nexus/blobstore/restore/p2/internal/P2RestoreBlobStrategy.java [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+    return new P2RestoreBlobData(restoreBlobData);
+  }
+
+  @Override
+  protected boolean canAttemptRestore(@Nonnull final P2RestoreBlobData p2RestoreBlobData) {
+    Repository repository = getRepository(p2RestoreBlobData);
+    Optional<P2RestoreFacet> p2RestoreFacetFacet = repository.optionalFacet(P2RestoreFacet.class);
+
+    if (!p2RestoreFacetFacet.isPresent()) {
+      log.warn("Skipping as R Restore Facet not found on repository: {}", repository.getName());
```
Comment (proteus1121) [inline nexus-restore-p2/src/main/java/org/sonatype/nexus/blobstore/restore/p2/internal/P2RestoreBlobStrategy.java]: p2

File: nexus-repository-p2/src/main/java/org/sonatype/nexus/repository/p2/internal/P2FacetImpl.java (hunk 6 above)
Comment (sergeyonofreychuck) [inline nexus-repository-p2/src/main/java/org/sonatype/nexus/repository/p2/internal/P2FacetImpl.java]: we already have appropriate DTO P2Attributes. Why used a hashmap ?

Comment (sergeyonofreychuck) [reply nexus-repository-p2/src/main/java/org/sonatype/nexus/repository/p2/internal/P2FacetImpl.java]: We already have doCreateOrSaveComponent in P2ProxyFacetImpl. 
findOrCreateComponent and findOrCreateAsset looks like a code duplication

File: nexus-repository-p2/src/main/java/org/sonatype/nexus/repository/p2/internal/P2FacetImpl.java [hunk 10]
```diff
@@ -0,0 +1,283 @@
...
+   * @return content of asset blob
+   */
+  public Content toContent(final Asset asset, final Blob blob) {
+    Content content = new Content(new BlobPayload(blob, asset.requireContentType()));
+    Content.extractFromAsset(asset, HASH_ALGORITHMS, content.getAttributes());
+    return content;
+  }
+
+  @Override
+  public AssetKind getAssetKind(final String path) {
```
Comment (sergeyonofreychuck) [inline nexus-repository-p2/src/main/java/org/sonatype/nexus/repository/p2/internal/P2FacetImpl.java]: add unit tests tor this function.
@proteus1121  please recheck the logic of asset kind discovery

File: nexus-repository-p2/src/main/java/org/sonatype/nexus/repository/p2/internal/P2FacetImpl.java (hunk 6 above)
Comment (DreamVan) [reply nexus-repository-p2/src/main/java/org/sonatype/nexus/repository/p2/internal/P2FacetImpl.java]: P2ProxyFacet is not visible from the module, the same for P2Attributes. Also, doCreateOrSaveComponent have as parameter Payload which we don't have on this case

File: nexus-repository-p2/src/main/java/org/sonatype/nexus/repository/p2/internal/P2FacetImpl.java (hunk 10 above)
Comment (proteus1121) [reply nexus-repository-p2/src/main/java/org/sonatype/nexus/repository/p2/internal/P2FacetImpl.java]: looks same as in P2ProxyRecipe.configure()

File: nexus-repository-p2/src/main/java/org/sonatype/nexus/repository/p2/internal/P2RestoreFacetImpl.java [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+    return Query.builder().where(P_NAME).eq(attributes.getComponentName())
+        .and(P_VERSION).eq(attributes.getComponentVersion()).build();
+  }
+
+  private P2Attributes getComponentAttributes(final Blob blob, final String blobName, final String blobStoreName)
+      throws IOException
+  {
+    P2Attributes.Builder attributes = P2Attributes.builder();
+    AssetKind assetKind = facet(P2Facet.class).getAssetKind(blobName);
+    if (AssetKind.COMPONENT_BINARY == assetKind) {
```
Comment (sergeyonofreychuck) [inline nexus-repository-p2/src/main/java/org/sonatype/nexus/repository/p2/internal/P2RestoreFacetImpl.java]: this code is a good candidate to move to P2PathUtils

**Pull request merged**